app/components/form_components/autocomplete/sort.js

```diff
@@ -0,0 +1,48 @@
+// Returns an array of integers representing the position of each regex match.
+const matchPositions = (string, regexes) => regexes.map(regex => string.search(regex)).filter(i => i >= 0)
+
+const calculateWeight = (name, rawSynonyms, query) => {
+  const regexes = query.split(/\s+/).map(word => new RegExp('\\b' + word, 'i'))
+  const synonyms = rawSynonyms || []
+
+  const nameMatchPositions = matchPositions(name, regexes)
+  const synonymMatchPositions = synonyms.map(synonym => matchPositions(synonym, regexes)).flat()
+
+  // Require all parts of the query to be matched:
+  if (nameMatchPositions.length !== regexes.length && synonymMatchPositions.length !== regexes.length) return 0
+
+  // Case insensitive exact matches:
+  const nameIsExactMatch = query === name.toLowerCase()
+  const synonymIsExactMatch = synonyms.some(s => query === s.toLowerCase())
+
+  // Case insensitive 'starts with':
+  const nameStartsWithQuery = nameMatchPositions.includes(0)
+  const synonymStartsWithQuery = synonymMatchPositions.includes(0)
+  const wordInNameStartsWithQuery = nameMatchPositions.length > 0
+
+  if (nameIsExactMatch) return 100
+  if (synonymIsExactMatch) return 75
+  if (nameStartsWithQuery) return 60
+  if (synonymStartsWithQuery) return 50
+  if (wordInNameStartsWithQuery) return 25
+
+  return 0
+}
+
+const byWeightThenAlphabetically = (a, b) => {
+  if (a.weight > b.weight) return -1
+  if (a.weight < b.weight) return 1
+  if (a.name < b.name) return -1
+  if (a.name > b.name) return 1
+
+  return 0
+}
+
+export default (q, options) => {
+  const query = q.trim().replace(/[.,/#!$%^&*;:{}=\-_`~()]/g, '').toLowerCase()
```

Comment:
I wonder if this should be within `calculateWeight`? 

I think we could also have the same cleaning function applied to synonyms and names? If I'm reading correctly, if a synonym or name had punctuation, the user would still be required to type it to get a match.

Reply:
Good idea - vurrently adding tests for punctuation stripping from both sides 👍

Reply:
Done! Check the two last test files for functionality.